sql/core/src/main/scala/org/apache/spark/sql/execution/r/ArrowRRunner.scala

```diff
@@ -55,13 +56,32 @@ class ArrowRRunner(
     numPartitions = -1,
     isDataFrame = true,
     schema.fieldNames,
-    RRunnerModes.DATAFRAME_GAPPLY) {
+    mode) {
+
+  // TODO: it needs to refactor to share the same code with RRunner, and have separate
+  // ArrowRRunners.
+  private val getNextBatch = {
+    if (mode == RRunnerModes.DATAFRAME_GAPPLY) {
+      // gapply
+      (inputIterator: Iterator[_], keys: collection.mutable.ArrayBuffer[Array[Byte]]) => {
+        val (key, nextBatch) = inputIterator
+          .asInstanceOf[Iterator[(Array[Byte], Iterator[InternalRow])]].next()
+        keys.append(key)
+        nextBatch
+      }
+    } else {
+      // dapply
+      (inputIterator: Iterator[_], keys: collection.mutable.ArrayBuffer[Array[Byte]]) => {
+        inputIterator
+          .asInstanceOf[Iterator[Iterator[InternalRow]]].next()
+      }
+    }
+  }
 
   protected override def writeData(
       dataOut: DataOutputStream,
       printOut: PrintStream,
-      iter: Iterator[_]): Unit = if (iter.hasNext) {
-    val inputIterator = iter.asInstanceOf[Iterator[(Array[Byte], Iterator[InternalRow])]]
+      inputIterator: Iterator[_]): Unit = if (inputIterator.hasNext) {
```

Comment:
Is it possible for `inputIterator.hasNext` to be false, like from an empty partition, and is there a test for this? It might be better to send an empty Arrow stream in this case, but I'm not entirely sure what the R worker is expecting.

Reply:
Oh, yea. I actually manually tested them. I'll fix the tests to check it as well.

Actually, I noticed R worker works a bit differently.

If the iterator is empty, it sends `0` first and then the other data is not used. (see `isEmpty <- SparkR:::readInt(inputCon)` at `worker.R` and,

```scala
          if (!iter.hasNext) {
            dataOut.writeInt(0)
          } else {
            dataOut.writeInt(1)
          }
```

at `RRunner.scala`).

So, virtually it doesn't matter if we send the data or not. I just simply decided to ignore this since it's already ignored at R's worker side.

Reply:
and .. I target to add more type specification test (for instance, map type and struct type are not supported).

I tried to only switch data transfer logic so checking empty rows, non-empty rows, columns, type specifications might be okay enough. (BTW, existing dapply itself is a bit flaky with some holes IMHO so it's a bit tricky to compare both).

Reply:
Sounds good to me, thanks!